hal/explicit_pinmap.h

```diff
@@ -0,0 +1,183 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018-2019 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef EXPLICIT_PINMAP_H
+#define EXPLICIT_PINMAP_H
+#define CONSTEXPR constexpr
+#include "PeripheralPinMaps.h"
+
+constexpr PinMap get_pwm_pinmap_t(const PinName pin)
+{
+    for (uint32_t i = 0; i < (sizeof(PINMAP_PWM)/sizeof(PINMAP_PWM[0])); i++) {
+        if (PINMAP_PWM[i].pin == pin) {
+            return {pin, PINMAP_PWM[i].peripheral, PINMAP_PWM[i].function};
+        }
+    }
+    return {NC, NC, NC};
+}
+
+constexpr PinMap get_analogin_pinmap_t(const PinName pin)
+{
+    for (uint32_t i = 0; i < (sizeof(PINMAP_ANALOGIN)/sizeof(PINMAP_ANALOGIN[0])); i++) {
```

Comment:
Now we're in C++14 land, this can be

    for (const PinMap &pinmap: PINMAP_ANALOGIN) {
        if (pinmap.pin == pin) {
             return pinmap;
        }
    }

Or you could go nuts and do this:

    PinMap *pinmap= std::find_if(std::begin(PINMAP_ANALOGIN), std::end(PINMAP_ANALOGIN), 
                                  [&](const PinMap &e) { return e.pin == pin });
    if (pinmap) {
        return *pinmap;
    }

But that latter more-verbose form only makes sense if you were really thinking `PINMAP_ANALOGIN` might be some sort of container which wanted to do a faster-than-linear search for performance reasons, so forget I even mentioned it...

Reply:
Fixed, but only for peripherals with one pin. It seems that when more pin tables need to be seach and the results compared the current version is better.